agent/internal/fluent_intg_test.go

```diff
@@ -0,0 +1,257 @@
+	e := echo.New()
+	baseListener, err := net.Listen("tcp", fmt.Sprintf(":%d", 8080))
+	assert.NilError(t, err, "error starting mock master listener")
+	e.Listener = baseListener
+	logBuffer := make(chan model.TrialLog, 100)
+	e.POST("/trial_logs", func(ctx echo.Context) error {
+		body, err := ioutil.ReadAll(ctx.Request().Body)
+		if err != nil {
+			return err
+		}
+
+		var logs []model.TrialLog
+		if err = json.Unmarshal(body, &logs); err != nil {
+			return err
+		}
+
+		for _, l := range logs {
+			end := strings.Contains(*l.Log, "END")
+			logBuffer <- l
+			if end {
+				close(logBuffer)
+				break
+			}
+		}
+		return nil
+	})
+	go func() {
+		if err := e.StartServer(e.Server); err != nil {
+			t.Logf("mock master exited: %s", err)
+		}
+	}()
+	return logBuffer, func() {
+		if err := e.Close(); err != nil {
+			t.Logf("error closing mock master echo server: %s", err)
+		}
+		if err := e.Listener.Close(); err != nil {
+			t.Logf("error closing mock master listener: %s", err)
+		}
+	}
+}
+
+func runContainerWithLogs(t *testing.T, fakeLogs string, trialID, fluentPort int) {
+	// WHEN we start a container that logs with fluentbit as its log driver.
+	docker, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.40"))
+	assert.NilError(t, err, "error connecting to Docker daemon")
+
+	logs, err := docker.ImagePull(context.Background(), "busybox", types.ImagePullOptions{})
+	assert.NilError(t, err, "failed to pull image")
+	_, err = ioutil.ReadAll(logs)
+	assert.NilError(t, err, "failed to drain pull logs")
+
+	// AND set some known ENV values.
+	spec := cproto.Spec{
+		RunSpec: cproto.RunSpec{
+			ContainerConfig: container.Config{
+				Cmd:   []string{"cat", "/fakelogs"},
+				Image: "busybox",
+			},
+			UseFluentLogging: true,
+		},
+	}
+	env := []string{fmt.Sprintf("DET_TRIAL_ID=%d", trialID)}
+	cont := cproto.Container{ID: "goodcontainer"}
+	spec = overwriteSpec(cont, spec, env, nil, fluentPort)
+	cc, err := docker.ContainerCreate(
+		context.Background(),
+		&spec.RunSpec.ContainerConfig,
+		&spec.RunSpec.HostConfig,
+		&spec.RunSpec.NetworkingConfig,
+		"log-test-"+uuid.New().String())
+	assert.NilError(t, err, "error creating container")
+
+	files := []archive.Item{
+		{
+			Path:     "/fakelogs",
+			Type:     tar.TypeReg,
+			FileMode: 0444,
+			Content:  []byte(fakeLogs),
+		},
+	}
+	filesReader, err := archive.ToIOReader(files)
+	assert.NilError(t, err, "failed make reader from fluent files")
+
+	err = docker.CopyToContainer(context.Background(),
+		cc.ID,
+		"/",
+		filesReader,
+		types.CopyToContainerOptions{},
+	)
+	assert.NilError(t, err, "failed to copy files to container")
+
+	err = docker.ContainerStart(context.Background(), cc.ID, types.ContainerStartOptions{})
+	assert.NilError(t, err, "error starting container")
+}
+
+func assertLogEquals(t *testing.T, l, expected model.TrialLog) {
+	assert.Equal(t, l.TrialID, expected.TrialID, spew.Sdump(l))
+	assert.Assert(t, l.AgentID != nil, spew.Sdump(l))
```

Comment:
Can you use `DeepEqual` to skip the nil checks?

Reply:
I looked into this but then the error message for `assert.Assert(t, reflect.DeepEqual(...))` just says something like "deep equal was false" in all failure cases, whereas this way you get a detailed failure. I did peek at something like https://github.com/go-test/deep but it felt like overkill.